### PR TITLE
Findby joined lookup

### LINKED ISSUE #7512: findBy broken for inherited entity property lookups
### Bug Report

| Q.                 |   A
|------------ | ------
| BC Break     | no
| Version        | 2.6.3

#### Summary

findBy does not work correctly for inherited entities. proof here: https://github.com/themizzi/doctrine-find-by-metadata-bug

#### Current behavior

If EntityB has a OneToMany relationship to EntityA, and EntityC inherits (via join) from EntityA. If there is no cache, calling `$entityBRepo->findBy('entityA' => $newEntityCInstance)` works fine. On a subsequent request after cache has already been generated and metadata is lazy loaded, the request breaks. on sqlite this just fails, on MySQL this potentially loads up an incorrect entity.

#### How to reproduce

Please see link above for an example and reproduction steps.

#### Expected behavior

Find by can find classes by properties with an inheritance structure at all times.

### PR DESCRIPTION
**Pull request by beberlei**
Supersedes https://github.com/doctrine/orm/issues/7514

Problem with object parameters not recongized as entities when they are.

Fixes #7512

### CODE REVIEW

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 1]
```diff
@@ -1980,11 +1968,29 @@ private function getValues($value)
      */
     private function getIndividualValue($value)
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1969]: The return type seems to have changed, it should be like this now I think:

```
@return       array<mixed>
@psalm-return list<mixed>
```

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 2]
```diff
@@ -1980,11 +1968,29 @@ private function getValues($value)
...
     {
-        if ( ! is_object($value) || ! $this->em->getMetadataFactory()->hasMetadataFor(ClassUtils::getClass($value))) {
-            return $value;
+        if ( ! is_object($value)) {
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1971]: I think git-phpcs is indicating the wrong line to cs2pr. This should fix the cs issue.

```suggestion
        if (! is_object($value)) {
```

**Review by greg0ire — DISMISSED** [review]

**Review by greg0ire — DISMISSED** [review]

File: lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php [hunk 3]
```diff
@@ -1955,36 +1956,43 @@ private function getValues($value)
...
      * Retrieves an individual parameter value.
      *
      * @param mixed $value
      *
-     * @return mixed
+     * @return       array<mixed>
+     *
+     * @psalm-return list<mixed>
      */
     private function getIndividualValue($value)
```
Comment (ostrolucky) [inline lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1971]: I think method name is no longer correct. It was used only as a fallback when it was clear there is only one value

Comment (beberlei) [reply lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php:1971]: I am not sure, it still returns the parameters for an individual value (which can be an entity with a composite id), whereas the method "above" also handles the case of arrays of individual values.

**Review by ostrolucky — DISMISSED** [review]

**Review by SenseException — DISMISSED** [review]

**Pull request merged**